src/disco/downloader.js

```diff
@@ -0,0 +1,209 @@
+import { pipeline } from 'node:stream';
+import { createWriteStream } from 'node:fs';
+import https from 'node:https';
+
+
+
+
+ const DISCO_BATCH_URLS = [
+  "https://zenodo.org/records/14159931/files/batch_1.tar.gz?download=1",
+  "https://zenodo.org/records/14160154/files/batch_2.tar.gz?download=1",
+  "https://zenodo.org/records/14160213/files/batch_3.tar.gz?download=1",
+  "https://zenodo.org/records/14160221/files/batch_4.tar.gz?download=1",
+  "https://zenodo.org/records/14160748/files/batch_5.tar.gz?download=1",
+  "https://zenodo.org/records/14160802/files/batch_6.tar.gz?download=1",
+  "https://zenodo.org/records/14166702/files/batch_7.tar.gz?download=1",
+  "https://zenodo.org/records/15236185/files/batch_8.tar.gz?download=1",
+  "https://zenodo.org/records/15236615/files/batch_9.tar.gz?download=1"
+];
+
+const DISCO_METADATA_URL = "https://disco.bii.a-star.edu.sg/disco_v3_api/toolkit/getSampleMetadata";
+
+const DISCO_BASE_URL = "https://disco.bii.a-star.edu.sg/sample/"
+
+
+const ORGAN_MAPPING = {
+
+  // Add more mappings as needed
+};
+
+const execFile = promisify(callbackExecFile);
+const streamPipeline = promisify(pipeline);
+
+
+/** @implements {IDownloader} */
+export class Downloader {
+  constructor(config) {
+    /** @type {Config} */
+    this.config = config;
+    /** @type {string} */
+    this.cacheDir = getCacheDir(config);
+    /** @type {string} */
+    this.baseUrl = DISCO_BASE_URL; 
+    /** @type {string[]} */
+    this.batchFiles = DISCO_BATCH_URLS.map(url => {
+      const batchName = url.match(/batch_\d+\.tar\.gz/)[0];
+      return join(this.cacheDir, batchName);
+    });
+    /** @type {string} */
+    this.extractScriptFile = 'extract_dataset.py';
+    /** @type {string} */
+    this.metadataFile = "disco_metadata.tsv";
+        /** @type {string} */
+    this.metadataFilePath = join(this.cacheDir, this.metadataFile);
+    /** @type {string} */
+    this.extractScriptFilePath = getSrcFilePath(
+      config,
+      'disco',
+      this.extractScriptFile
+    );
+    /** @type {OrganMetadataCollection} */
+    this.organMetadata = undefined;
+  }
+
+  async prepareDownload(datasets) {
+
+    this.organMetadata = await OrganMetadataCollection.load(this.config);
+      await downloadFile(this.metadataFilePath, DISCO_METADATA_URL, {
+        overwrite: this.config.get(FORCE, false)
+      });
+
+    // Download all batch tar files
+
+    await Promise.all(DISCO_BATCH_URLS.map(async (url) =>  {
+
+    // Extract batch name 
+    const batchName = url.match(/batch_\d+\.tar\.gz/)[0].replace('.tar.gz', '');
+    const expectedExtractedDir = join(this.cacheDir, batchName);
+
+    // Skip if already extracted
+    if (existsSync(expectedExtractedDir) && !this.config.get(FORCE, false)) {
+      console.log(`Skipping already extracted batch: ${batchName}`);
+      return;
+    }
+
+    console.log(`Streaming and extracting: ${url}`);
+
+    await new Promise((resolve, reject) => {
+      https.get(url, (res) => {
```

Comment:
use execFile like this: https://github.com/hubmapconsortium/hra-workflows-runner/blob/main/src/ts2/downloader.js#L54-L55 to run command line so

```javascript
const targetDir = join(this.cacheDir, batchName);
fs.mkdirSync(targetDir, { recursive: true });  // ensure target dir exists
await execFile(`curl ${url} | tar xf --strip-components=1 -C ${targetDir}`);
```